### PR TITLE
Tests for RSC10 & RSC14d

### PR DESCRIPTION
**Pull request by adamgiacomelli**
Inital attempt on these tests - still checking if everything is ok, would love some comments.

### CODE REVIEW

**Comment by paddybyers** [comment]
@SimonWoolf can you have a look pls?

**Comment by SimonWoolf** [comment]
:+1: looks good to me as far as it goes.

Would be nice to have a test for the second half of RSC14d as well, as that's the part that I'm not sure is implemented ("then the client should automatically make a single attempt to reissue the token and resend the request using the new token. If the token creation failed or the subsequent request with the new token failed due to a token error, then the request should result in an error")

eg you could have a test with authCallback that always returns the same token it got originally, and check that it's only called one additional time after it expires when the client lib errors. (And maybe another test that does the exact same thing but with an authCallback that returns just the token string rather than tokenDetails, so the client lib can't see that it's expired).

Mind adding those?

**Comment by adamgiacomelli** [comment]
@SimonWoolf Will do, thanks for the review and the example

**Comment by adamgiacomelli** [comment]
@mattheworiordan @SimonWoolf added the first test for the single retry, seems the functionality is missing - if you could take a look it would be lovely. I have to check the code to see if its implemented.

File: spec/rest/auth.test.js [hunk 1]
```diff
@@ -615,5 +615,57 @@ define(['ably', 'shared_helper', 'async'], function(Ably, helper, async) {
...
+		var single_token = null;
+		var request_counter = 0;
+
+		var staticTokenCallback = function(tokenParams, callback) {
+			tokenParams.ttl = 1;
+			//Always return the same token
+			if(single_token == null) {
+				rest.auth.requestToken(tokenParams, function(err, tokenDetails) {
+					if(err) {
+						console.log("Token error");
```
Comment (mattheworiordan) [inline spec/rest/auth.test.js]: Is this needed given the error is going to be displayed by assertion below?

**Comment by mattheworiordan** [comment]
>  added the first test for the single retry, seems the functionality is missing - if you could take a look it would be lovely.

What error are you getting @adamgiacomelli?

File: spec/rest/auth.test.js (hunk 1 above)
Comment (adamgiacomelli) [inline spec/rest/auth.test.js]: The debug printout was a leftover from debugging - I have removed it .

**Comment by adamgiacomelli** [comment]
@mattheworiordan I am not getting an error at all - the token callback is getting called in a loop - I just stop the test after it gets called more than once after its reissued (per @SimonWoolf 's example) .

**Comment by mattheworiordan** [comment]
ok thanks @adamgiacomelli, we'll get that fixed then and see if the tests passes.

**Comment by SimonWoolf** [comment]
@adamgiacomelli added a fix (https://github.com/ably/ably-js/commit/3190ec175f5444ef645e41ab7d86fd78166864cb) to the `reauthOnlyOnce` branch, could you rebase on that and retest, please.

I believe to make it pass you'll have to make staticTokenCallback call back with `single_token.token` rather than `single_token` (if auth has the full tokenDetails then getAuthHeaders notices that the token has expired and takes the opportunity to request a new one, which throws the counting off. Technically that's not quite spec-compliant, but not in a way I really care about - whether it's once or twice doesn't matter that much, main point is that it doesn't do an infinite loop, which was the previous behaviour).

I also included a blind fix for the realtime equivalent (RTN14b and RTN15h): https://github.com/ably/ably-js/commit/5ac9c70d96a1113454a3fea2bc7cdf7dd4f853f6. Haven't tested. Could you please write a couple test for the single-attempt part of those specs to make sure the fix is correct, checking that they fail without that commit and pass with it. Thanks :)

**Comment by adamgiacomelli** [comment]
@SimonWoolf tested the reauthOnlyOnce branch and did the necessary modifications - the test passes easily, will get to the realtime equivalents shortly

**Pull request closed without merging**